programs/Makefile

```diff
@@ -52,6 +52,10 @@ CPPFLAGS  += -I$(ZSTDDIR)/legacy
 ZSTDLEGACY_FILES:= $(ZSTDDIR)/legacy/*.c
 endif
 
+ifeq ($(shell uname), FreeBSD)
+CPPFLAGS  += -D_POSIX_C_SOURCE=200112L
```

Comment:
1. The issue concerns the availability of `-r` (recursive) option that uses `opendir/closedir` functions (http://man.he.net/man3/closedir) conforming to SVr4, POSIX.1-2001, 4.3BSD

2. The `_POSIX_C_SOURCE=200112L` macro is used to detect availability of `opendir/closedir` functions on Linux. My understanding is that `_POSIX_C_SOURCE` shouldn't be defined by user at all because it may not work e.g. with older versions of an operating system and `glibc`. It's better to use `--std=gnu99` or `-D_GNU_SOURCE` which should be more portable as they define a proper version `_POSIX_C_SOURCE`.

3. The proposed changes force the C parser to turn on '-r' on FreeBSD. But it's better to do auto detection of FreeBSD only for relevant file that is `util.h`. It seems that all BSD-forks are based on 4.4BSD therefore we need only to check availability of one of:
__DragonFly__
__FreeBSD__
__NetBSD__
__OpenBSD__